### PR TITLE
ddl: avoid panic when creating partition table with unsupported expression

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
TiDB will panic when create partition table using unsupported expression.
```
mysql> create table t1 (a int) partition by range (a + (select * from t)) (partition p1 values less than (1));
ERROR 2013 (HY000): Lost connection to MySQL server during query
```
This is because `buildTablePartitionInfo` use `ExprNode.Format` method to convert expression to string. Some `ExprNode` like `SubQueryExpr` will panic when use their `Format` method.

```
// Format the ExprNode into a Writer.
func (n *SubqueryExpr) Format(w io.Writer) {
	panic("Not implemented")
}
```

### What is changed and how it works?

- Add check whether the partition expression is valid before executing `buildTablePartitionInfo`
- Make check rules stronger.

### Check List 

Tests 

 - Unit test

Code changes

 - None

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

**Comment by imtbkcat** [comment]
PTAL @jackysp @crazycs520

**Comment by jackysp** [comment]
PTAL @crazycs520

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1382,9 +1382,16 @@ func buildTableInfoWithCheck(ctx sessionctx.Context, s *ast.CreateTableStmt, dbC
 		return nil, errors.Trace(err)
 	}
 
-	pi, err := buildTablePartitionInfo(ctx, s)
-	if err != nil {
-		return nil, errors.Trace(err)
+	var pi *model.PartitionInfo
+	if s.Partition != nil {
```
Comment (wjhuang2016) [inline ddl/ddl_api.go]: Please move L1385 and `L1397 to L1411` into this scope

File: ddl/partition.go [hunk 2]
```diff
@@ -294,10 +294,9 @@ func defaultTimezoneDependent(ctx sessionctx.Context, tblInfo *model.TableInfo,
 	return !v, nil
 }
 
-// checkPartitionFuncValid checks partition function validly.
-func checkPartitionFuncValid(ctx sessionctx.Context, tblInfo *model.TableInfo, expr ast.ExprNode) error {
+func checkPartitionExprValid(ctx sessionctx.Context, tblInfo *model.TableInfo, expr ast.ExprNode) error {
```
Comment (wjhuang2016) [inline ddl/partition.go]: Please add the comment

**Comment by imtbkcat** [comment]
PTAL @wjhuang2016

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1382,25 +1382,31 @@ func buildTableInfoWithCheck(ctx sessionctx.Context, s *ast.CreateTableStmt, dbC
...
-	}
-
-	if pi != nil {
-		switch pi.Type {
-		case model.PartitionTypeRange:
-			err = checkPartitionByRange(ctx, tbInfo, pi, cols, s)
-		case model.PartitionTypeHash:
-			err = checkPartitionByHash(ctx, pi, s, cols, tbInfo)
-		}
+	var pi *model.PartitionInfo
```
Comment (wjhuang2016) [inline ddl/ddl_api.go]: This line can be removed

**Pull request merged**